profiler.go

```diff
@@ -0,0 +1,221 @@
+package sentry
+
+import (
+	"runtime"
+	"strconv"
+	"time"
+
+	"github.com/getsentry/sentry-go/internal/traceparser"
+)
+
+// Start collecting profile data and returns a function that stops profiling, producing a Trace.
+// May return nil or an incomplete trace in case of a panic.
+func startProfiling() func() *profilerResult {
+	// buffered channels to handle the recover() case without blocking
+	result := make(chan *profilerResult, 2)
+	stopSignal := make(chan struct{}, 2)
+
+	go profilerGoroutine(result, stopSignal)
+
+	return func() *profilerResult {
+		stopSignal <- struct{}{}
+		return <-result
+	}
+}
+
+// This allows us to test whether panic during profiling are handled correctly and don't block execution.
+var testProfilerPanic = 0
+var testProfilerPanickedWith any
+
+func profilerGoroutine(result chan<- *profilerResult, stopSignal chan struct{}) {
+	// We shouldn't panic but let's be super safe.
+	defer func() {
+		testProfilerPanickedWith = recover()
+		// Make sure we don't block the caller of stopFn() even if we panic.
+		result <- nil
+	}()
+
+	// Stop after 30 seconds unless stopped manually.
+	timeout := time.AfterFunc(30*time.Second, func() { stopSignal <- struct{}{} })
+	defer timeout.Stop()
+
+	if testProfilerPanic == -1 {
+		testProfilerPanic = 0
+		panic("This is an expected panic in profilerGoroutine() during tests")
+	}
+
+	profiler := newProfiler()
+
+	// Collect the first sample immediately.
+	profiler.onTick()
+
+	// Periodically collect stacks, starting after profilerSamplingRate has passed.
+	collectTicker := time.NewTicker(profilerSamplingRate)
+	defer collectTicker.Stop()
+
+	defer func() {
+		result <- &profilerResult{profiler.startTime, profiler.trace}
+	}()
+
+	for {
+		select {
+		case <-collectTicker.C:
+			profiler.onTick()
+		case <-stopSignal:
+			return
+		}
+	}
+}
+
+type profilerResult struct {
+	startTime time.Time
+	trace     *profileTrace
+}
+
+func newProfiler() *profileRecorder {
+	// Pre-allocate the profile trace for the currently active number of routines & 100 ms worth of samples.
+	// Other coefficients are just guesses of what might be a good starting point to avoid allocs on short runs.
+	numRoutines := runtime.NumGoroutine()
+	trace := &profileTrace{
+		Frames:         make([]*Frame, 0, 32),
+		Samples:        make([]*profileSample, 0, numRoutines*10), // 100 ms @ 101 Hz
+		Stacks:         make([]profileStack, 0, 8),
+		ThreadMetadata: make(map[string]profileThreadMetadata, numRoutines),
+	}
+
+	return &profileRecorder{
+		startTime:    time.Now(),
```

Comment:
I agree and I've originally had just a copy of the transaction timestamp here but with the current implementation starting the profiler in a goroutine, it may happen some time later after the transaction starts so using the same timestamp would give incorrect info. This issue will be obsolete if https://github.com/getsentry/sentry-go/pull/626#discussion_r1203659275 is resolved

Reply:
It wouldn't. It would actually be good. The important of this timestamp is not to tell us when the profiler actually started, the timing of the first sample can tell us that.

What we're doing with this timestamp is accepting a timestamp when we can't possibly pass the transaction start time (for example, when we don't control the profiler). Here, we do control it so it'd be better to pass the start timestamp of the transaction.

It doesn't actually matter if the profiler starts profiling 100ms later than this timestamp as the goal is to be then able to have every span and sample aligned on the same start time instead of having to calculate that based on that timestamp.

Reply:
ok then, changed